src/CityOnPlanet.cpp

```diff
@@ -347,25 +679,32 @@ void CityOnPlanet::Render(Graphics::Renderer *r, const Graphics::Frustum &frustu
 	}
 
 	// update any idle animations
-	for (Uint32 i = 0; i < s_buildingList.numBuildings; i++) {
-		SceneGraph::Animation *pAnim = s_buildingList.buildings[i].idle;
-		if (pAnim) {
-			pAnim->SetProgress(fmod(pAnim->GetProgress() + (Pi::game->GetTimeStep() / pAnim->GetDuration()), 1.0));
-			pAnim->Interpolate();
+	// TODO: this is kind of a horrible idea in many ways
+	// animated buildings need per-instance state and need to support instanced drawing with GPU animation transform of some kind
+	// individual cities at the *very* least need to have their own copy of the model used in them so color + animation state
+	// doesn't "leak" into other users of the model
+	for (auto &buildingType : m_cityType->buildingTypes) {
+		SceneGraph::Animation *anim = buildingType.idleAnimation;
+		if (anim) {
+			anim->SetProgress(fmod(anim->GetProgress() + (Pi::game->GetTimeStep() / anim->GetDuration()), 1.0));
+			anim->Interpolate();
 		}
 	}
 
-	Uint32 uCount = 0;
-	std::vector<Uint32> instCount;
+	uint32_t uCount = 0;
+	uint32_t numBuildings = m_cityType->buildingTypes.size();
+
+	std::vector<uint32_t> instCount;
 	std::vector<std::vector<matrix4x4f>> transform;
-	instCount.resize(s_buildingList.numBuildings);
-	transform.resize(s_buildingList.numBuildings);
-	memset(&instCount[0], 0, sizeof(Uint32) * s_buildingList.numBuildings);
-	for (Uint32 i = 0; i < s_buildingList.numBuildings; i++) {
+
+	instCount.resize(numBuildings, 0);
+	transform.resize(numBuildings);
+
+	for (uint32_t i = 0; i < numBuildings; i++) {
 		transform[i].reserve(m_buildingCounts[i]);
 	}
 
-	for (std::vector<BuildingDef>::const_iterator iter = m_enabledBuildings.begin(), itEND = m_enabledBuildings.end(); iter != itEND; ++iter) {
+	for (std::vector<BuildingInstance>::const_iterator iter = m_enabledBuildings.begin(), itEND = m_enabledBuildings.end(); iter != itEND; ++iter) {
```

Comment:
It seems that nothing prevents us from using range-based for here, and not all that.

Reply:
Indeed, I didn't (yet) do a full pass on the code to convert it all to a new style. Will do that before merge.